prober/grpc.go

```diff
@@ -0,0 +1,55 @@
+package prober
+
+import (
+	"context"
+	"strings"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/prometheus/blackbox_exporter/config"
+
+	"google.golang.org/grpc"
+	healthpb "google.golang.org/grpc/health/grpc_health_v1"
+)
+
+func ProbeGRPC(ctx context.Context, target string, module config.Module, registry *prometheus.Registry, logger log.Logger) bool {
+	var (
+		conn    *grpc.ClientConn
+		parts   []string
+		service string
+		host    string
+	)
+
+	parts          = strings.Split(target, "/")
+	host, service  = parts[0], parts[1]
+	conn, err     := grpc.Dial(host, grpc.WithInsecure())
+
+	if err != nil {
+		level.Error(logger).Log("msg", "failed to connect")
+		return false
+	}
+
+	defer conn.Close()
+
+	level.Error(logger).Log("msg", "invoking", "host", host, "server", service, "method", module.GRPC.Method, "request", healthpb.HealthCheckRequest{Service: service})
+
+	req := &healthpb.HealthCheckRequest{Service: service}
+	res := new(healthpb.HealthCheckResponse)
+
+	err = grpc.Invoke(
+		ctx,
+		module.GRPC.Method,
+		req,
+		res,
+		conn,
+	)
+
+	if err != nil {
+		level.Error(logger).Log("msg", err)
+		return false
+	}
+
+	return true
```

Comment:
The result is a `HealthCheckResponse` which contains a `ServingStatus` field (https://github.com/grpc/grpc/blob/e5dc51fcf371df7e79e62efd05881d7c7f0a2fe1/src/proto/grpc/health/v1/health.proto#L39) which may indicate unhealthy (i.e. anything but `SERVING`). I think this should also return false if the response is not `SERVING`.